### DISCUSSION TITLE
Issue with Implementing  `ToSql` and ` FromSql` for ` UserRoleMapping` in Diesel

### DISCUSSION THREAD

**Discussion by sony-level** [diesel-rs/diesel#4086] [Q&A] [answered]
**Description:**
Hello,

I am encountering a problem while compiling my project using Diesel for the implementation of ToSql and FromSql functions. The purpose of these functions is to convert the UserRole enum to a textual value for storing in a database and to read textual values from the database and convert them to UserRole.

Purpose of the Implemented Functions:

- ToSql: This function converts an instance of UserRoleMapping to a textual value that can be stored in a database.

- FromSql: This function converts a textual value from the database to an instance of UserRoleMapping.

**Errors Encountered:**
When running cargo build, I receive the following error:

```
`error[E0599]: no method named `write_all` found for mutable reference `&mut diesel::serialize::Output<'b, '_, DB>` in the current scope
  --> src/models.rs:78:13
   |
78 |         out.write_all(self.0.as_bytes())?;
   |             ^^^^^^^^^ method not found in `&mut Output<'b, '_, DB>`

`
```

**Versions Used:**

Rust: 1.26.0
Diesel: 2.2.1

**Source Code:**
```

`#[derive(Debug, Serialize, Deserialize)]
pub enum UserRole {
    Trainer,
    Student,
}

/**
 * UserRole enum
 * Enum for user roles
 * trainer: trainer
 * student: student
 * This implementation is necessary to convert the UserRole enum to a textual value that can be stored in the database. 
 * Diesel uses this implementation to insert and update UserRole values in Text columns. 
 * Diesel also uses this implementation to read values from the database and convert them to UserRole values.
 */
impl fmt::Display for UserRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            UserRole::Trainer => write!(f, "trainer"),
            UserRole::Student => write!(f, "student"),
        }
    }
}

impl From<&str> for UserRole {
    fn from(s: &str) -> Self {
        match s {
            "trainer" => UserRole::Trainer,
            "student" => UserRole::Student,
            _ => panic!("Invalid value for UserRole"),
        }
    }
}

#[derive(AsExpression, FromSqlRow, Debug)]
#[diesel(sql_type = Text)]
struct UserRoleMapping(String);

impl From<UserRole> for UserRoleMapping {
    fn from(role: UserRole) -> Self {
        UserRoleMapping(role.to_string())
    }
}

impl From<UserRoleMapping> for UserRole {
    fn from(mapping: UserRoleMapping) -> Self {
        UserRole::from(mapping.0.as_str())
    }
}

impl<DB> ToSql<Text, DB> for UserRoleMapping
where
    DB: Backend,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> serialize::Result {
        out.write_all(self.0.as_bytes())?;
        Ok(serialize::IsNull::No)
    }
}

impl FromSql<Text, Pg> for UserRoleMapping {
    fn from_sql(value: PgValue<'_>) -> deserialize::Result<Self> {
        let s = std::str::from_utf8(value.as_bytes())?;
        Ok(UserRoleMapping(s.to_string()))
    }
}

impl FromSql<Text, Pg> for UserRole {
    fn from_sql(value: PgValue<'_>) -> deserialize::Result<Self> {
        let s = std::str::from_utf8(value.as_bytes())?;
        Ok(UserRole::from(s))
    }
}
`

```
I cannot understand why the method `write_all` is not found in ` &mut Output<'b, '_, DB>`. Any help or guidance to resolve this issue would be greatly appreciated.

Thank you in advance for your assistance.

**Comment by weiznich** [comment, marked as answer]
This is expected behaviour as the `Write` trait is not implemented for all backends as shown in the [documentation](https://docs.diesel.rs/2.2.x/diesel/serialize/struct.Output.html#impl-Write-for-Output%3C'a,+'b,+DB%3E). You either need to use a non generic backend type for your `ToSql` implementation or you need to add this bound there as well or you need to write the `ToSql` implementation in terms of existing implementations. 

The documentation of [`ToSql`](https://docs.diesel.rs/2.2.x/diesel/serialize/trait.ToSql.html) has a few details about these variants.

**Reply by sony-level** [reply]
Hi @weiznich ,

Thank you very much for your response and the detailed explanation. With your help, I understood that the issue was due to using a generic `backend` for the `ToSql` implementation. To address this, I explored an alternative approach using the `diesel_derive_enum` crate, which provides a simpler and more elegant solution for handling enum types in Diesel.

Here’s how I modified my code:

```
use diesel_derive_enum::DbEnum;

#[derive(Debug, Serialize, Deserialize, DbEnum)]
pub enum UserRole {
    #[db_rename = "trainer"]
    Trainer,
    #[db_rename = "student"]
    Student,
}

#[derive(AsExpression, FromSqlRow, Debug)]
#[diesel(sql_type = Text)]
pub struct UserRoleMapping(UserRole);

```

With this change, the `ToSql` and ` FromSql` functions are automatically derived for UserRole, greatly simplifying the code and avoiding the compilation errors previously encountered.

Thanks again for your valuable assistance!

Best regards,

Sony
